File: medium/1680/main.c
#define MOD 1000000007

int concatenatedBinary(int n){
  long long res = 0;

  for (int i = 1; i <= n; i++) {
    int cnt = 0;
    int tmp = i;
    while (tmp) {
      cnt++;
      tmp >>= 1;
    }

    res = ((res << cnt) + i) % MOD;
  }

  return (int) res;

}

File: medium/1680/main.rs
impl Solution {
    pub fn concatenated_binary(n: i32) -> i32 {
        let mut res: i64 = 0;
        let mut len: i64 = 0;
        let MOD: i64 = 1_000_000_007;

        for i in 1..=n {
            if (i & (i-1)) == 0 {
                len += 1;
            }
            res = ((res << len) | (i as i64)) % MOD;
        }
        
        res as i32
    }
}
